src/views/Browse/Browser.tsx

```diff
@@ -0,0 +1,100 @@
+import { useCallback, useMemo } from "react";
+import { useLocation, useNavigate, useSearchParams } from "react-router-dom";
+
+import { ApiBrowseRes } from "../../../src/common/types";
+import { Spinner } from "../../components";
+import { joinUrl, setOrDeleteParam } from "../../utils";
+import { Filters, HeaderCell, Navigation, Preview, Row, RowBack, Sorting } from "./components";
+import { useFilter, useSort } from "./providers";
+
+export const Browser: React.FC<{
+  contents: ApiBrowseRes | undefined;
+  loading: boolean;
+  path: string;
+}> = ({ contents, loading, path }) => {
+  const location = useLocation();
+  const navigate = useNavigate();
+  const [params, setParams] = useSearchParams();
+
+  const { applyFilters } = useFilter();
+  const { applySorting } = useSort();
+
+  const browseAbsolute = useCallback(
+    (url: string) => {
+      // preseve sorting
+      const nextParams = new URLSearchParams();
+      setOrDeleteParam(nextParams, "sort", params.get("sort") ?? undefined);
+      setOrDeleteParam(nextParams, "dirs", params.get("dirs") ?? undefined);
+
+      navigate(url + `?${nextParams.toString()}`);
+    },
+    [navigate, params],
+  );
+
+  const browseRelative = useCallback(
+    (name: string) => browseAbsolute(joinUrl("/", location.pathname, name)),
+    [location.pathname, browseAbsolute],
+  );
+
+  const preview = useMemo(() => params.get("preview"), [params]);
+  const showPreview = useCallback(
+    (value: string) => void setParams(setOrDeleteParam(params, "preview", value)),
+    [params, setParams],
+  );
```

Comment:
Thanks for giving me a nudge to go through the usage of `useCallback` in detail! [I've kept it only in places](https://github.com/tadeaspetak/teleport-challenge/pull/5/commits/1df68dabeac2553e958aff42b469b49f3040820a) where:

1. The function is used as a dependency for another hooks such as `useMemo` and `useEffect`, which is the majority of the cases.
2. The function is passed down to child components to avoid unnecessary re-renders (such as in the [Browser](https://github.com/tadeaspetak/teleport-challenge/blob/1df68dabeac2553e958aff42b469b49f3040820a/src/views/Browse/Browser.tsx#L34)).

Reply:
I have read the articles and have experimented with component re-rendering and memoization. My understanding of components inside contexts was flawed: I thought that when you destructure a context such as `const { toast } = useToast()`, the component would update only when that particular object (`toast` here) of the context updates. Naturally, this is far from truth — unless memoized, the component, as any child, will update every time its parent, here the context, re-renders.

I've gone through the code, removed a fair amount of `useMemo` and `useCallback` calls and left comments on why I believe they should be used wherever they remained [in this commit](https://github.com/tadeaspetak/teleport-challenge/pull/5/commits/18066fd55ecec2c6be4ec78fe77434ea5c6f7169).

When optimising, I was focused on "I don't want the `Browser` to update when an unrelated  context is updated (e.g. `toast` messages)" and, naturally, on "I want filters & sorting to be automatically synced with the state kept in the URL."

Next optimisation would be making sure that `Row` doesn't re-render when URL params (preview, filtering, sorting) change. I thought that should be easy, but I haven't been able to figure out how to set search params in a functional way with ReactRouter v6. I thought there would be a function signature similar to `setState(prev => ...)`, but `setSearchParams` needs `searchParams`, and since the `Row`s need a way to navigate/show preview, they now re-render on each change of the parameters. I will dig deeper into this.

Reply:
Here's my best take on making the URL `params` stable:

https://github.com/tadeaspetak/teleport-challenge/blob/5ee8d77a49fc7b416c8933e8d00b80679ac78a2e/src/hooks/useDeferredParams.ts#L4-L15

... and here it's usage in the `Browser`:

https://github.com/tadeaspetak/teleport-challenge/blob/5ee8d77a49fc7b416c8933e8d00b80679ac78a2e/src/views/Browse/Browser.tsx#L25-L45

When the `Row` is memoized with `React.memo`, it now doesn't unnecessarily re-render on sorting and filter changes. Hurray!